Allows for viewing old readonly versions

pipwilson opened the pull request:
Reads the value of planVersion from sessionService and makes a request through a new planService method if planVersion exists.

Also adds a small bit of extra data integrity maintenance by checking the plan.readOnly property when rendering the Plan Overview and redirecting to the read-only view if so. This is so that users visiting old plan versions with READ_WRITE role can't make edits.

pipwilson committed:
Allows for viewing old readonly versions

Thomas-Geraghty commented on server/@types/PlanType.ts:
```diff
@@ -28,4 +28,5 @@ export type PlanType = {
   updatedBy: string
   agreementDate: string
   goals: Goal[]
+  readOnly: boolean
```
What's the need for this? Isn't the `read-only` attribute determined by the `READ_ONLY` role coming from OASys  (maintained in the handover context)

pipwilson replied:
Our spec for PlanType does have a readOnly value and I thought this was a good point at which to introduce it. If we're not going to use it we should probably take it out of our object type in the API as well.

pipwilson replied:
removed in https://github.com/ministryofjustice/hmpps-sentence-plan-ui/pull/206/commits/057bb2e02687440e511407401ee7829f85c2667b

pipwilson committed:
Removes readOnly attribute on PlanType

pipwilson committed:
Adds command for locking plan and initial countersigning e2e test using different plan versions

pipwilson committed:
Improve access mode handling (#207)

This was by thomas.geraghty <32871802+Thomas-Geraghty@users.noreply.github.com>

pipwilson committed:
Updates linting and tidies tests but still failing

Thomas-Geraghty committed:
Fix unit tests

Thomas-Geraghty committed:
Fix lint

Thomas-Geraghty committed:
Fix typecheck

pipwilson committed:
change undefined reference to null

pipwilson committed:
quick patch for countersign test

alexw-moj commented on integration_tests/e2e/countersign.cy.ts:
```diff
@@ -26,3 +28,33 @@ describe('View Plan Overview for READ_ONLY user', () => {
+    })
+  })
+
+  it('Should have one button', () => {
```
Think there's more work to do here, tests nothing about buttons.

pipwilson replied:
finished this as much as I can right now in https://github.com/ministryofjustice/hmpps-sentence-plan-ui/pull/206/commits/5d7fe1e6e2b9300569a9695c5e767ccf1dfc8554 - ideally I think it would sign out and then sign back in again to check the list twice but our sign-out doesn't quite work right so trusting that `addGoalToPlan` is I think as much as I can do now!

alexw-moj commented on server/middleware/authorisationMiddleware.ts:
```diff
@@ -1,21 +1,30 @@
   }
 }
+
+export function hasAccessMode(requiredAccessMode: AccessMode): RequestHandler {
```
Can we rename this to `validateAccessMode`?

pipwilson replied:
renamed in https://github.com/ministryofjustice/hmpps-sentence-plan-ui/pull/206/commits/db72907e836bee0b129166ca960242c987f07028

pipwilson committed:
adds more to the countersign test

pipwilson committed:
rename middleware accessmode check method
